File: helperFunctions/sumById.py
import listToDict
import json

# some functions to help with logging
from helperFunctions.sendMessages import printParent
from helperFunctions.sendMessages import messageParent
from helperFunctions.sendMessages import obviousPrint


def sum( dataDescription, X, headerRow, idColumn, trainingLength, outputColumn):
    if checkForDupes(idColumn, trainingLength):
        printParent('We have found multiple rows with the same ID in them.')
        printParent('We are going to assume that all rows with the same ID in them should be summed up intelligently')
        printParent('This will tranform your dataset from being "long" to being "wide".')
        printParent('If this is not what you intended, please submit an issue and/or a Pull Request explaining your sitaution!')
        # TODO: we now have two different return formats: dictionaries, and lists
        # probably easiest to convert X to dictionaries here regardless of whether it has dupes or not
        # hmmm, imputing missing values would likely be less useful for these cases
            # it's more likely that we would have a separate file entirely for the metadata associated with each row (name, age, gender if our repeated ID is a customerID)
            # let's ignore this for now (MVP!), and then think later about imputing missing values on only the non-joined data, then joining in that data later. that would likely be much more space efficient than joining in that data up front
        return groupByID(dataDescription, X, headerRow, idColumn, trainingLength, outputColumn)
        

    else:
        return [listToDict.all( X, headerRow ), idColumn, trainingLength, outputColumn]

# check to see if we have duplicate IDs in this data set
def checkForDupes( idColumn, trainingLength ):
    idCounts = {}
    for rowIndex, ID in enumerate(idColumn):
        try: 
            if idCounts[ID] == 1 and rowIndex < trainingLength:
                # if we can access this property in idCounts, that means we already have a value there, and can return true!
                return True
        except:
            idCounts[ID] = 1

    return False

def groupByID(dataDescription, X, headerRow, idColumn, trainingLength, outputColumn):
    # TODO: take in the output column as well, add it to each rowObj, and then split back out again following the exact same process we are for the idColumn
    # FUTURE: handle data where the IDs are not sorted
        # We could easily just save this into a giant dictionary, instead of a results list
        # The key would be the ID, and the value would be this rowObj
        # Then we could just create it if it doesn't exist, and add to it if it does exist


    # FUTURE: support multiple continuous values for each row (number of items purchased, and total $ sales, for example)
    valueIndex = dataDescription.index('continuous')
    valueHeader = headerRow[ valueIndex ]

    results = {}
    trainingIDs = {}
    newTrainingLength = 0
    # iterate through list
    for rowIndex, row in enumerate(X):
        rowID = idColumn[rowIndex]

        # create a rowObj for each rowID, if we don't have one for this rowID already saved in our results dictionary
        try:
            rowObj = results[rowID]
            rowObj['rowCount'] += 1
        except:
            results[rowID] = {}
            rowObj = results[rowID]
            # the number of rows this ID will appear in
            rowObj['rowCount'] = 1
            # the number of different categories this row will hold overall
            rowObj['categoryCount'] = 0
            # this ensures we will always have the ID value attached to the summarized results for this row, even after we turn the results dictionary into a list
            # we take this column back out again after dictVectorizer
            rowObj['id'] = rowID

            # if we have to create a new rowObj, that means that we have not encountered this ID before. 
            # not encountering this ID before, and being in a position within our X dataset that is less than the training length, means that this is a new, unique row summary that belongs to our trainingLength

            if rowIndex < trainingLength:
                newTrainingLength += 1
                trainingIDs[rowID] = True
                rowObj['output'] = outputColumn[rowIndex]
        # There is going to be one value for each row (e.g. number of items sold)
        rowValue = row[ valueIndex ]
        for columnIndex, value in enumerate(row):
            # Then, for each categorical value (department, sub-department, etc.), sum up that rowValue
            # Thus, if we make multiple purchases in the biking section, we will have all of that summed up for this ID (either customer or trip or anything else we are given as an ID)
            if dataDescription[columnIndex] == 'categorical':
                columnHeader = headerRow[ columnIndex ]
                try:
                    rowObj[ columnHeader + value + valueHeader ] += rowValue
                    # how many different times this ID has had this value for this column
                        # e.g., how many times the sub-department "Road Bikes" has popped up for this ID
                    rowObj[ columnHeader + value + valueHeader + 'count' ] += rowValue
                except:
                    # for each column that is categorical, add a property to the rowObj, if it does not exist already 
                    # add the continuous value for that row to the value for this property 
                    rowObj[ columnHeader + str(value) + valueHeader ] = rowValue
                    rowObj[ columnHeader + str(value) + valueHeader + 'count' ] = rowValue
                    # total categories this ID will eventually hold
                    rowObj['categoryCount'] += 1

                    # total categories- for this column- this ID will eventually hold
                        # e.g., if we have a column dedicated to the sub-departments of that store, how many different sub-departments (Road Bikes, Mountain Bikes, and Rock Climbing) this ID will hold
                    try:
                        rowObj[columnHeader + 'count'] += 1
                    except:
                        rowObj[columnHeader + 'count'] = 1



                # FUTURE: support having multiple continuous values added for each ID (count and price, for example)
                # create counts of all these variables
                    # number of rows
                    # number of rows for each categorical value (dairy, for example)

    listResults = results.values()
    idColumn = []

    # TODO: 
        # create a list of IDs that are in the training set
        # once we have turned back into a list again, iterate through each object
        # based on it's ID, put it into either a training or testing dataset
        # then at the end, concat the training and testing datasets together again. 
        # 

    # since we have no idea what order the rowDicts will be in once we've put them into listResults, we must go through and carefully separate out the training and testing datasets
    # part of that separation includes creating a new idColumn, since we now have fewer rows, and a new outputColumn, for the same reason.
    trainingData = []
    trainingIDColumn = []
    testingData = []
    testingIDColumn = []
    outputColumn = []
    for rowDict in listResults:
        rowID = rowDict.pop('id', None)
        try:
            if trainingIDs[str(rowID)] == True:
                trainingData.append(rowDict)
                trainingIDColumn.append( rowID )
                outputColumn.append( rowDict.pop('output', None) )
            else:
                testingData.append( rowDict )
                testingIDColumn.append( rowID )
        except: 
            testingData.append( rowDict )
            testingIDColumn.append( rowID )
    
    listResults = trainingData + testingData
    idColumn = trainingIDColumn + testingIDColumn
    return [listResults, idColumn, newTrainingLength, outputColumn]

File: testSuite/wmt/sumById.js
var expect = require('chai').expect;
var mocha = require('mocha');

module.exports = function() {

  describe('sumByID', function() {
    
    var idColumn;
    var outputColumn;
    var X;
    var trainingLength;

    before(function(done) {

      console.time('sumByID time');
      pyControllerWmt.on('message', function(message) {
        if(message.type === 'sumById.py') {

          idColumn = message.text[1];
          trainingLength = message.text[2];
          outputColumn = message.text[3];
          X = message.text[0];

          console.timeEnd('sumByID time');
          done();
        }
      });
    });

    it('should turn each ID into a single row of data', function() {
      var idSummaryObj = {};

      function summarizeIDColumn() {
        for (var i = 0; i < idColumn.length; i++) {
          var id = idColumn[i];
          if( idSummaryObj[id] !== undefined) {
            return false;
          } else {
            idSummaryObj[id] = 1;
          }
        }
        return true;
      }

      expect( summarizeIDColumn( idColumn) ).to.be.true;

    });

    it('should not have id or output column in the rowObj itself, only in their own separate columns', function() {
      function checkForIDAndOutput(results) {
        for( var i = 0; i < results.length; i++) {
          var rowObj = results[i];
          for( var key in rowObj ) {
            if( key === 'id' || key === 'output') {
              console.log('rowObj:',rowObj);
              return false;
            }
          }
        }
        return true;
      };

      expect( checkForIDAndOutput(X) ).to.be.true;
    });



    it('should return a list of dictionaries', function() {
      function checkForObjects(results) {
        for( var i = 0; i < results.length; i++) {
          if( results[i].constructor !== Object) {
            return false;
          }
        }
        return true;
      }

      expect( checkForObjects(X) ).to.be.true;

    });

    it('should have columns for each category', function() {

      function countColumns(results) {
        var count = 0;
        var columnSummary = {};
        for( var i = 0; i < results.length; i++) {
          for( var key in results[i]) {
            if( columnSummary[key] === undefined) {
              count++;
              columnSummary[key] = true;
            } 
          }
        }
        return count;
      }
      expect( countColumns(X) ).to.equal(660);
    });

    // it('should sum up results by id for each category', function() {
    //   // TODO: get the expected sum of each aggregated column by calculating it manually
    //     // compare the observed sums to those to make sure it's calculating correctly!
    // });


    it('should keep track of which IDs are part of the training set, and which are in the testing set', function() {
      var trainingIDs = [5,7,8,9,10,11,12,15,17,19,20,23,25,26,28,29,30,31,32,33,40,41,42,43,45,47,49,50,51,53,54,55,56,57,61,63,68,69,72,74,76,79,81,83,84,85,86,87,91,92,93,95,97,98,99,100,102,103,105,106,107,110,117,121,123,124,125,130,132,133,134,137,138,139,140,145,146,152,153,154,155,162,164,177,179,181,182,185,188,189,190,194,195,199,203,205,207,209,210,211,214,216,217,218,219,220,221,223,224,225,228,229,232,235,236,238,239,241,245,248,251,252,253,255,256,257,259,261,263,265,267,274,275,277,278,279,281,283,284,285,286,287,288,289,290,291,292,295,298,302,308,309,310,312,313,314,315,316,317,322,323,324,325,326,328,330,333,334,335,341,343,344,346,347,348,349,350,351,357,358,359,360,361,362,363,367,369,371,372,373,375,377,378,379,382,383,384,385,386,387,388,389,390,393,394,396,397,398,399,400,402,403,407,408,409,412,413,418,419,420,421,425,427];
      var testingIDs = [1,2,3,4,6,13,14,16,18,21,22,24,27,34,35,36,37,38,39,44,46,48,52,58,59,60,62,64,65,66,67,70,71,73,75,77,78,80,82,88,89,90,94,96,101,104,108,109,111,112,113,114,115,116,118,119,120,122,126,127,128,129,131,135,136,141,142,143,144,147,148,149,150,151,156,157,158,159,160,161,163,165,166,167,168,169,170,171,172,173,174,175,176,178,180,183,184,186,187,191,192,193,196,197,198,200,201,202,204,206,208,212,213,215,222,226,227,230,231,233,234,237,240,242,243,244,246,247,249,250,254,258,260,262,264,266,268,269,270,271,272,273,276,280,282,293,294,296,297,299,300,301,303,304,305,306,307,311,318,319,320,321,327,329,331,332,336,337,338,339,340,342,345,352,353,354,355,356,364,365,366,368,370,374,376,380,381,391,392,395,401,404,405,406,410,411,414,415,416,417,422,423,424,426,428,429,432,434,435,438,439,440,441,442,443,444,445,450,451,454,455,459,463,464,467];
      function checkIDs(idColumn, trainingLength) {
        for( var i = 0; i < trainingLength; i++) {
          if( trainingIDs.indexOf( parseInt(idColumn[i], 10) ) === -1) {
            console.log('the id that is incorrectly sorted as a training ID is:',idColumn[i]);
            return idColumn[i];
          }
        }
        for (var i = trainingLength; i < idColumn.length; i++) {
          if( testingIDs.indexOf( parseInt(idColumn[i], 10) ) === -1) {
            console.log('the id that is incorrectly sorted as a testing ID is:',idColumn[i]);
            return idColumn[i];
          }
        }
        return true;
      }
      expect( checkIDs(idColumn, trainingLength) ).to.be.true;

    });

    // it('should return new idColumn and outputColumn lists that match up to the order of X', function() {
    //   // check some IDs, and test against their known output value
    //   // also check lengths of these lists
    // });

    after(function() {
      idColumn = null;
      outputColumn = null;
      X = null;
      trainingLength = null;
    });
    
  });

};
